### PR TITLE
[FIRRTL][HW] Change default implementation of CombDataFlow and add it to DPI intrinsic 

### PR DESCRIPTION
**Pull request by uenoku**
This PR removes default implementation of CombDataFlow which indicate an op was sequential. Also implement it for DPI intrinsic.

Close https://github.com/llvm/circt/issues/7229

### CODE REVIEW

**Review by darthscsi — COMMENTED** [review]
It's an alternate implementation, sure.  It's not obviously faster.  Control flow and indirect calls are quite expensive, possibly more so than dense storage passed by value.

**Comment by uenoku** [comment]
> It's an alternate implementation, sure. It's not obviously faster. Control flow and indirect calls are quite expensive, possibly more so than dense storage passed by value.

I don't have a strong opinion TBH. CombDataFlow interface is opt-in and not heavily used right now (only used by MemoryOp) so it would be hard to do microbenchmark

File: include/circt/Dialect/HW/HWOpInterfaces.td [hunk 1]
```diff
@@ -556,17 +556,17 @@ def CombDataflow : OpInterface<"CombDataFlow"> {
...
-     "the dependence."
-     "The default implementation returns an empty list, which implies that the"
-     "operation is not combinational.",
-     "SmallVector<std::pair<circt::FieldRef, circt::FieldRef>>",
-     "computeDataFlow", (ins), [{}], /*defaultImplementation=*/[{
-      return {};
-    }]>,
+    InterfaceMethod<[{
+     Get the combinational dataflow relations between the operands and the results.
+     This returns a pair of ground type fieldrefs. The first element is the destination
```
Comment (prithayan) [inline include/circt/Dialect/HW/HWOpInterfaces.td:561]:
```suggestion
     The callback is invoked with a pair of ground type fieldrefs. The first element is the destination
```
The comment needs to be updated as it no longer returns it, something along the lines of the suggestion.

**Review by prithayan — APPROVED** [review]
The updated interface looks good to me.

**Comment by uenoku** [comment]
I reverted a change for using a callback for now. I'll follow up if necessary.

**Pull request merged**